File: photo-studio/tests/test_zip_handler.py
from __future__ import annotations

from pathlib import Path

import pytest

from app.zip_handler import (
    ZipValidationError,
    detect_product_folders,
    safe_extract_zip,
    validate_zip_bytes,
)


def test_safe_extract_extracts_only_valid_images(tmp_path, sample_zip_path, test_settings):
    dest = tmp_path / "extracted"
    report = safe_extract_zip(sample_zip_path, dest, test_settings)

    # 2 + 3 + 1 + 1 = 7 valid images
    assert report.extracted_files == 7
    assert (dest / "ITEM_01" / "photo1.jpg").exists()
    assert (dest / "ITEM_02" / "label.png").exists()
    assert (dest / "ITEM_04" / "photo.webp").exists()


def test_path_traversal_is_neutralised(tmp_path, sample_zip_path, test_settings):
    dest = tmp_path / "extracted"
    report = safe_extract_zip(sample_zip_path, dest, test_settings)

    # The "../../evil.jpg" entry must never escape the destination directory.
    escaped_path = tmp_path / "evil.jpg"
    assert not escaped_path.exists()
    assert not (tmp_path.parent / "evil.jpg").exists()
    assert any("evil.jpg" in w for w in report.warnings)

    # Every file that *was* written must be inside dest.
    for path in dest.rglob("*"):
        if path.is_file():
            assert dest.resolve() in path.resolve().parents


def test_ignores_macosx_and_dsstore_and_non_images(tmp_path, sample_zip_path, test_settings):
    dest = tmp_path / "extracted"
    safe_extract_zip(sample_zip_path, dest, test_settings)

    assert not (dest / "__MACOSX").exists()
    assert not (dest / ".DS_Store").exists()
    assert not (dest / "ITEM_01" / "notes.txt").exists()


def test_detect_product_folders_groups_correctly(tmp_path, sample_zip_path, test_settings):
    dest = tmp_path / "extracted"
    safe_extract_zip(sample_zip_path, dest, test_settings)
    products, warnings = detect_product_folders(dest, test_settings)

    names = sorted(p.name for p in products)
    assert names == ["ITEM_01", "ITEM_02", "ITEM_03", "ITEM_04"]

    by_name = {p.name: p for p in products}
    assert len(by_name["ITEM_01"].images) == 2
    assert len(by_name["ITEM_02"].images) == 3
    assert len(by_name["ITEM_03"].images) == 1
    assert len(by_name["ITEM_04"].images) == 1


def test_validate_zip_bytes_rejects_oversized(test_settings):
    too_big = (test_settings.max_zip_size_mb + 1) * 1024 * 1024
    with pytest.raises(ZipValidationError):
        validate_zip_bytes(too_big, test_settings)


def test_validate_zip_bytes_rejects_empty(test_settings):
    with pytest.raises(ZipValidationError):
        validate_zip_bytes(0, test_settings)


def test_max_product_folders_limit_enforced(tmp_path, sample_zip_path, test_settings):
    dest = tmp_path / "extracted"
    safe_extract_zip(sample_zip_path, dest, test_settings)

    from dataclasses import replace

    strict_settings = replace(test_settings, max_product_folders=2)
    with pytest.raises(ZipValidationError):
        detect_product_folders(dest, strict_settings)


def test_corrupt_zip_raises(tmp_path, test_settings):
    bad_zip = tmp_path / "bad.zip"
    bad_zip.write_bytes(b"this is not a zip file at all")
    with pytest.raises(ZipValidationError):
        safe_extract_zip(bad_zip, tmp_path / "out", test_settings)


def test_no_images_in_zip_raises(tmp_path, test_settings):
    import zipfile

    empty_zip = tmp_path / "empty.zip"
    with zipfile.ZipFile(empty_zip, "w") as zf:
        zf.writestr("ITEM_01/notes.txt", b"no images here")
    with pytest.raises(ZipValidationError):
        safe_extract_zip(empty_zip, tmp_path / "out", test_settings)

File: photo-studio-php/tests/test_zip_handler.php
<?php
declare(strict_types=1);

test('safe_extract extracts only valid images', function () {
    $tmp = make_temp_dir();
    try {
        $zipPath = $tmp . '/sample.zip';
        make_sample_zip($zipPath);
        $dest = $tmp . '/extracted';
        $report = jps_safe_extract_zip($zipPath, $dest);

        assert_equals(7, $report['extracted'], 'expected 7 valid images extracted');
        assert_true(is_file($dest . '/ITEM_01/photo1.jpg'));
        assert_true(is_file($dest . '/ITEM_02/label.png'));
        assert_true(is_file($dest . '/ITEM_04/photo.webp'));
    } finally {
        rrmdir($tmp);
    }
});

test('path traversal is neutralised', function () {
    $tmp = make_temp_dir();
    try {
        $zipPath = $tmp . '/sample.zip';
        make_sample_zip($zipPath);
        $dest = $tmp . '/extracted';
        $report = jps_safe_extract_zip($zipPath, $dest);

        assert_false(is_file($tmp . '/evil.jpg'), 'evil.jpg must not escape the extraction dir');
        assert_true(count(array_filter($report['warnings'], fn($w) => str_contains($w, 'evil.jpg'))) > 0);

        $destReal = realpath($dest);
        $rii = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($dest, FilesystemIterator::SKIP_DOTS));
        foreach ($rii as $file) {
            if ($file->isFile()) {
                assert_true(str_starts_with(realpath($file->getPathname()), $destReal), 'file escaped dest dir: ' . $file->getPathname());
            }
        }
    } finally {
        rrmdir($tmp);
    }
});

test('ignores __MACOSX, .DS_Store and non-images', function () {
    $tmp = make_temp_dir();
    try {
        $zipPath = $tmp . '/sample.zip';
        make_sample_zip($zipPath);
        $dest = $tmp . '/extracted';
        jps_safe_extract_zip($zipPath, $dest);

        assert_false(is_dir($dest . '/__MACOSX'));
        assert_false(is_file($dest . '/.DS_Store'));
        assert_false(is_file($dest . '/ITEM_01/notes.txt'));
    } finally {
        rrmdir($tmp);
    }
});

test('detect_product_folders groups correctly', function () {
    $tmp = make_temp_dir();
    try {
        $zipPath = $tmp . '/sample.zip';
        make_sample_zip($zipPath);
        $dest = $tmp . '/extracted';
        jps_safe_extract_zip($zipPath, $dest);
        [$products, $warnings] = jps_detect_product_folders($dest);

        $names = array_map(fn($p) => $p['name'], $products);
        sort($names);
        assert_equals(['ITEM_01', 'ITEM_02', 'ITEM_03', 'ITEM_04'], $names);

        $byName = [];
        foreach ($products as $p) {
            $byName[$p['name']] = $p;
        }
        assert_equals(2, count($byName['ITEM_01']['images']));
        assert_equals(3, count($byName['ITEM_02']['images']));
        assert_equals(1, count($byName['ITEM_03']['images']));
        assert_equals(1, count($byName['ITEM_04']['images']));
    } finally {
        rrmdir($tmp);
    }
});

test('validate_zip_size rejects oversized and empty', function () {
    assert_throws(fn() => jps_validate_zip_size((MAX_ZIP_SIZE_MB + 1) * 1024 * 1024), JpsZipException::class, 'oversized zip must throw');
    assert_throws(fn() => jps_validate_zip_size(0), JpsZipException::class, 'empty zip must throw');
});

test('corrupt zip raises', function () {
    $tmp = make_temp_dir();
    try {
        $badZip = $tmp . '/bad.zip';
        file_put_contents($badZip, 'this is not a zip file at all');
        assert_throws(fn() => jps_safe_extract_zip($badZip, $tmp . '/out'), JpsZipException::class);
    } finally {
        rrmdir($tmp);
    }
});

test('no images in zip raises', function () {
    $tmp = make_temp_dir();
    try {
        $zipPath = $tmp . '/empty.zip';
        $zip = new ZipArchive();
        $zip->open($zipPath, ZipArchive::CREATE);
        $zip->addFromString('ITEM_01/notes.txt', 'no images here');
        $zip->close();
        assert_throws(fn() => jps_safe_extract_zip($zipPath, $tmp . '/out'), JpsZipException::class);
    } finally {
        rrmdir($tmp);
    }
});
